### ISSUE TITLE
Constants can be redefined without Warning using the 'def' Keyword

### ISSUE THREAD

**Issue by maflAT** [microsoft/pyright#7440]
<h3> Environment data </h3>

-   Language Server version: 2024.2.106
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda): 3.12
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: standard

<h3> Code Snippet </h3>

```python
from typing import Final

def FOO(): ...
FOO = lambda: None

BAR = lambda: None
def BAR(): ...

baz: Final = lambda: None
def baz(): ...
```

<h3> Expected behavior </h3>

Warnings 'reportConstantRedefinition' are shown for all 3 functions `FOO`, `BAR` and `baz`.

<h3> Actual behavior </h3>

![grafik](https://github.com/microsoft/pyright/assets/13979550/e0a1d5e1-41d7-4695-bd6a-f3ad23f2d376)

The warning only shows when the `=` assignment comes after the `def` assignment. 
If the `def` assignment comes second, there is no warning.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.354, which I just published.
